gnmi_server/server.go

```diff
@@ -515,3 +527,63 @@ func (s *Server) Capabilities(ctx context.Context, req *gnmipb.CapabilityRequest
 		GNMIVersion:        "0.7.0",
 		Extension:          exts}, nil
 }
+
+type uint128 struct {
+	High uint64
+	Low  uint64
+}
+
+func (lh *uint128) Compare(rh *uint128) int {
+	if rh == nil {
+		// For MA disabled case, EID supposed to be 0.
+		rh = &uint128{High: 0, Low: 0}
+	}
+	if lh.High > rh.High {
+		return 1
+	}
+	if lh.High < rh.High {
+		return -1
+	}
+	if lh.Low > rh.Low {
+		return 1
+	}
+	if lh.Low < rh.Low {
+		return -1
+	}
+	return 0
+}
+
+// ReqFromMasterEnabledMA returns true if the request is sent by the master
+// controller.
+func ReqFromMasterEnabledMA(req *gnmipb.SetRequest, masterEID *uint128) bool {
+	// Read the election_id.
+	reqEID := uint128{High: 0, Low: 0}
+	// It can be one of many extensions, so iterate through them to find it.
+	for _, e := range req.GetExtension() {
+		ma := e.GetMasterArbitration()
+		if ma == nil {
+			continue
+		}
+		// The Master Arbitration descriptor has been found.
+		// If MA is enabled, the reqEID needs to be set to 0
+		if ma.ElectionId != nil {
+			reqEID = uint128{High: ma.ElectionId.High, Low: ma.ElectionId.Low}
+		}
+		// Use the election ID that is in the last extension, so, no 'break' here.
```

Comment:
Is it better to iterate backwards to pick the last value?

Reply:
Done

Reply:
Since we have checked is `ma == nil` previously, the last available MA extension should have its EID updated here.